### PR TITLE
[IOTDB-4627]Trigger transfer

### PR DESCRIPTION
**Pull request by Wei-hao-Li**
1.Transfer stateful triggers when DataNode Status is Unknown;
2.Transfer stateful triggers when DataNode remove;
3.Has been tested, can see [Trigger测试方案](https://apache-iotdb.feishu.cn/docx/doxcn9uRj5bpzb58MqgeMokjMC0)
4.Cooperate with @lancelly.

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java [hunk 1]
```diff
@@ -255,8 +255,17 @@ public DataSet removeDataNode(RemoveDataNodePlan removeDataNodePlan) {
...
+
     DataNodeToStatusResp dataSet = new DataNodeToStatusResp();
+    // do transfer of the DataNodes before remove
+    if (configManager.transfer(removeDataNodePlan.getDataNodeLocations()).getCode()
+        != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+      dataSet.setStatus(
+          new TSStatus(TSStatusCode.NODE_DELETE_FAILED_ERROR.getStatusCode())
+              .setMessage("Fail to do transfer of the DataNodes"));
+      return dataSet;
+    }
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java:267]: Is the `transfer` process should be such important that the DataNode remove process couldn't continue as long as if it fails down?

**Review by CRZbulabula — COMMENTED** [review]
PTAL～

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java (hunk 1 above)
Comment (Wei-hao-Li) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java:267]: Yes, I think we should ensure the safety of data.

**Review by CRZbulabula — APPROVED** [review]
LGTM

**Review by JackieTien97 — APPROVED** [review]

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/TriggerManager.java [hunk 2]
```diff
@@ -132,8 +146,79 @@ public TGetTriggerJarResp getTriggerJar(TGetTriggerJarReq req) {
...
+    } finally {
+      triggerInfo.releaseTriggerTableLock();
+    }
+
     return new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
   }
+
+  public List<TSStatus> updateTriggerLocation(
+      String triggerName,
+      TDataNodeLocation dataNodeLocation,
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/TriggerManager.java:213]: parameter `dataNodeLocation` and `dataNodeLocationMap` are similar, maybe we can change `dataNodeLocation` to `targetDataNode`?

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java (hunk 1 above)
Comment (Beyyes) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java:267]: Remove data process used RemoveDataProcedure and sub RegionMigrateProcedures, 
Will triggers implement it later or using other methods?
Or some errors may happen when ConfigNode down/restart.

image

Comment (Wei-hao-Li) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java:267]: We used to consider implement by Procedure, but later we found that we did not need to roll back when the transfer failed.

**Pull request merged**